MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,578 @@
+        {
+            var pos = startPosition >= 0 ? startPosition : 0;
+            return new ReaderInfo(this, pos, 0, length, isMotorolaByteOrder);
+        }
+
+        /// <summary>Retrieves bytes, writing them into a caller-provided buffer.</summary>
+        /// <param name="index">position within the data buffer to read byte.</param>
+        /// <param name="buffer">array to write bytes to.</param>
+        /// <param name="offset">starting position within <paramref name="buffer"/> to write to.</param>
+        /// <param name="count">number of bytes to be written.</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <returns>The requested bytes, or as many as can be retrieved</returns>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        public int Read(long index, byte[] buffer, int offset, int count, bool isSequential)
+        {
+            return Read(index, buffer, offset, count, isSequential, true);
+        }
+
+        /// <summary>Retrieves bytes, writing them into a caller-provided buffer.</summary>
+        /// <param name="index">position within the data buffer to read byte.</param>
+        /// <param name="buffer">array to write bytes to.</param>
+        /// <param name="offset">starting position within <paramref name="buffer"/> to write to.</param>
+        /// <param name="count">number of bytes to be written.</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <param name="allowPartial">flag indicating whether count should be enforced when validating the index</param>
+        /// <returns>The requested bytes, or as many as can be retrieved if <paramref name="allowPartial"/> is true</returns>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        public int Read(long index, byte[] buffer, int offset, int count, bool isSequential, bool allowPartial)
+        {
+            count = (int)ValidateIndex(index, count, isSequential, allowPartial);
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            // TODO: good spot to try Span<T>
+            if (p_isStreamFinished && p_chunks.Count == 1)
+            {
+                Array.Copy(p_chunks[0], (int)index, buffer, 0, count);
+                return count;
+            }
+
+            var remaining = count;                      // how many bytes are requested
+            var fromOffset = (int)index;
+            var toIndex = offset > 0 ? offset : 0;
+            while (remaining != 0)
+            {
+                var fromChunkIndex = fromOffset / p_chunkLength;     // chunk integer key
+                var fromInnerIndex = fromOffset % p_chunkLength;     // index inside the chunk to start reading
```

Comment:
In the case where we read from multiple chunks, can we avoid the div/mod for all but the first iteration? Would it work if these were initialised before the loop, then at the end of the loop incremented and set to zero respectively?